document-policy-explainer.md

```diff
@@ -0,0 +1,341 @@
+just the content that it embeds. It can be used in conjunction with the
+`Document-Policy` header, though.
+
+When you set a required document policy, most features will have to be
+acknowledged by the document being embedded, as otherwise it can be possible to
+affect the content on that page in unexpected ways. Any policies which require
+acknowledgement will be advertised with a `Sec-Required-Document-Policy` request
+header.
+
+If the content which is returned doesn't have a `Document-Policy` header which
+is at least as strict as the `Sec-Required-Document-Policy`, then it will not be
+loaded into the frame, and a network error will result instead.
+
+The `Sec-Required-Document-Policy` header will be present for all requests for
+documents in a frame, at every nesting level. If the framed document adds its
+own `Require-Document-Policy header`, or uses the policy attribute on its own
+iframes, then the requirements will be combined in deeper requests, to ensure
+that all content meets all of the requirements being imposed on it.
+
+### Extending the Iframe Sandboxing Mechanism
+
+All existing sandbox features can be defined as document policies as well. The
+iframe `sandbox` attribute will disable all of them by default, but they can be
+turned back on either with the `policy` attribute or the `sandbox` attribute.
+Additionally, they can be disabled individually, even on frames which are not
+otherwise sandboxed.
+
+Any new sandbox-style features can be added through this mechanism, without
+having to add them to the set of features disabled by the sandbox attribute.
+
+### Resetting policies in popups
+
+There is currently a sandbox flag, the
+*sandbox propagates to auxiliary browsing contexts flag*, which controls whether
+the existing sandbox features should be inherited in popups opened from a
+sandboxed frame. This flag can continue to function as it currently does, and
+can be extended to cover the entire document policy. This flag itself, as a
+sandbox flag, can be controlled independently as well.
+
+### More Details: Opting in to being embedded with a policy
+
+Each feature has an associated flag, ***requires-acknowledgement***, which is
+true for all of those features which would be unsafe to impose on nested content
+without that content's approval (and is true by default if not explicitly
+defined).
+
+(Alternately, we can define a central safelist of features which do not require
+acknowledgement; the decision comes down to whether that call is made by the
+feature policy spec or the specs which define the individual features)
+
+As a starting point for discussion, all existing sandbox flags would not require
+acknowledgement, and all other document policies would.
+
+In documents generated from `data:` URLs, and in iframe srcdoc documents, where
+the parent document controls the content of the child explicitly, and there is
+no HTTP network transfer, the acknowledgment will simply be implied.
```

Comment:
How about same-origin iframes?

Reply:
That's a possibility -- does it open up any new kinds of attacks, where being able to affect the policy on an iframe and not requiring it to consent enables things you couldn't do otherwise?

Reply:
I'm not to sure if there are plausible attacks here, but it does seem like the same argument as for `data:` and `srcdoc` applies: the same origin controls both things.

This all reminds me of the decision to not require `<iframe allow="fullscreen">` for the same-origin case, although it's just analogous and risks might be different.